clients/src/main/java/org/apache/kafka/clients/consumer/internals/AbstractCoordinator.java

```diff
@@ -0,0 +1,638 @@
+     * on the preference).
+     * @return Non-empty map of supported protocols and metadata
+     */
+    protected abstract LinkedHashMap<String, ByteBuffer> metadata();
+
+    /**
+     * Invoked when a group member has successfully joined a group.
+     * @param generation The generation that was joined
+     * @param memberId The identifier for the local member in the group
+     * @param protocol The protocol selected by the coordinator
+     * @param memberAssignment The assignment propagated from the group leader
+     */
+    protected abstract void onJoin(int generation,
+                                   String memberId,
+                                   String protocol,
+                                   ByteBuffer memberAssignment);
+
+    /**
+     * Perform synchronization for the group. This is used by the leader to push state to all the members
+     * of the group (e.g. to push partition assignments in the case of the new consumer)
+     * @param leaderId The id of the leader (which is this member)
+     * @param allMemberMetadata Metadata from all members of the group
+     * @return A map from each member to their state assignment
+     */
+    protected abstract Map<String, ByteBuffer> doSync(String leaderId,
+                                                      String protocol,
+                                                      Map<String, ByteBuffer> allMemberMetadata);
+
+    /**
+     * Invoked when the group is left (whether because of shutdown, metadata change, stale generation, etc.)
+     * @param generation The generation that was left
+     * @param memberId The identifier of the local member in the group
+     */
+    protected abstract void onLeave(int generation, String memberId);
+
+
+    /**
+     * Block until the coordinator for this group is known.
+     */
+    public void ensureCoordinatorKnown() {
+        while (coordinatorUnknown()) {
+            RequestFuture<Void> future = sendGroupMetadataRequest();
+            client.poll(future, requestTimeoutMs);
+
+            if (future.failed())
+                client.awaitMetadataUpdate();
+        }
+    }
+
+    /**
+     * Check whether the group should be rejoined (e.g. if metadata changes)
+     * @return true if it should, false otherwise
+     */
+    protected boolean needRejoin() {
+        return rejoinNeeded;
+    }
+
+    /**
+     * Reset the generation/memberId tracked by this member
+     */
+    public void resetGeneration() {
+        this.generation = OffsetCommitRequest.DEFAULT_GENERATION_ID;
+        this.memberId = JoinGroupRequest.UNKNOWN_MEMBER_ID;
+        rejoinNeeded = true;
+    }
+
+    /**
+     * Ensure that the group is active (i.e. joined and synced)
+     */
+    public void ensureActiveGroup() {
+        if (!needRejoin())
+            return;
+
+        // onLeave only invoked if we have a valid current generation
+        onLeave(generation, memberId);
```

Comment:
@becketqin Thanks for the suggestions! A few notes:

1. I agree with renaming doSync. Initially I was thinking of this phase as "state synchronization," but in the end, it seemed that the assignment terminology was easier to understand, so I did some renaming, but didn't catch all. I think, however, that leaderId is needed in the general case because the member doesn't know its own id (the copycat patch is already depending on this).
2. Yeah, we can do this. Maybe more concisely, we can use `onBecomeLeader` and `onBecomeFollower`?
3. This suggestion is equally confusing in my mind. It seems more natural to phrase this in terms of the generation. Instead of `onLeave`, my suggestion would be `onGenerationEnd`.
4. Similarly, `onJoin` becomes `onGenerationBegin`.
5. I guess that's fair since it's not only sending the join group.